### PR TITLE
Chore: Fix data race within tests and enable a few parallel tests

### PR DESCRIPTION
**Pull request by diegommm**
**What is this feature?**

Fix a few data races in tests and enable running all tests in parallel in `pkg/services/ssosettings/ssosettingsimpl`.

**Why do we need this feature?**

To allow tests to run in parallel and enable the general use of `-race` for the grafana repo so we can automatically catch new data races introduced more easily.

**Who is this feature for?**

Everyone

**Which issue(s) does this PR fix?**:

None

**Special notes for your reviewer:**

Check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/ssosettings/ssosettingsimpl/service_test.go [hunk 1]
```diff
@@ -762,7 +798,7 @@ func TestService_Upsert(t *testing.T) {
 		reloadable := ssosettingstests.NewMockReloadable(t)
 		reloadable.On("Validate", mock.Anything, settings).Return(nil)
 		reloadable.On("Reload", mock.Anything, mock.MatchedBy(func(settings models.SSOSettings) bool {
-			wg.Done()
+			defer wg.Done()
 			return settings.Provider == provider &&
 				settings.ID == "someid" &&
 				maps.Equal(settings.Settings, map[string]any{
```
Comment (diegommm) [inline pkg/services/ssosettings/ssosettingsimpl/service_test.go:804]: The read of the map in `maps.Equal` was racing with a concurrent write after calling `wg.Wait()`, so I added the `defer` to make sure the read was performed before `wg.Wait()` and the subsequent write would be safe.

File: pkg/services/ssosettings/ssosettingsimpl/service_test.go [hunk 2]
```diff
@@ -794,17 +830,21 @@ func TestService_Upsert(t *testing.T) {
...
-		err := env.service.Upsert(context.Background(), &settings)
+		done, err := env.service.upsertAndReloadBg(context.Background(), &settings)
 		require.NoError(t, err)
+		require.NotNil(t, done)
 
 		// Wait for the goroutine first to assert the Reload call
 		wg.Wait()
+		<-done
 
 		settings.Settings["client_secret"] = base64.RawStdEncoding.EncodeToString([]byte("encrypted-client-secret"))
```
Comment (diegommm) [inline pkg/services/ssosettings/ssosettingsimpl/service_test.go:841]: `Upsert` and `Delete` being called here and then a subsequent write to the map were the cause of another set of race conditions. The former do work in a separate goroutine (reload) that was not synchronized with the write, so I made intermediate methods for them that return a done channel as well to allow synchronizing those operations with the write and later reads of the map within the tests, that are needed to assert the final state.

**Review by Jguer — COMMENTED** [review]
No strong feelings on this change. Happy parallel seems to now work in Grafana's code base. @mgyongyosi and @dmihai , up to you

File: pkg/services/ssosettings/ssosettingsimpl/service_test.go (hunk 1 above)
Comment (mgyongyosi) [reply pkg/services/ssosettings/ssosettingsimpl/service_test.go:804]: Thanks!

File: pkg/services/ssosettings/ssosettingsimpl/service.go [hunk 3]
```diff
@@ -196,9 +196,11 @@ func (s *Service) Upsert(ctx context.Context, settings *models.SSOSettings) erro
 		return err
 	}
 
-	settings.Settings = overrideMaps(storedSettings.Settings, settings.Settings, secrets)
+	// make a copy of current setings for reload operation and apply overrides
+	reloadSettings := *settings
+	reloadSettings.Settings = overrideMaps(storedSettings.Settings, settings.Settings, secrets)
 
-	go s.reload(social, settings.Provider, *settings)
+	go s.reload(social, settings.Provider, reloadSettings)
```
Comment (diegommm) [inline pkg/services/ssosettings/ssosettingsimpl/service.go:203]: Thanks @mgyongyosi for helping me realize here was the source of the other data race in the tests.

**Review by mgyongyosi — APPROVED** [review]
Thank you for the improvements! 🚀

**Pull request merged**